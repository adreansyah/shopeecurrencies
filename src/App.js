import React from 'react';
import { BrowserRouter as Router, Route, Switch } from 'react-router-dom';
import SiteExchange from 'pages/exchangecurrencies';

const App = prop => {
  return (
    <Router basename='/'>
      <Switch>
        <Route exact path='/' component={SiteExchange} />} />
        <Route exact path='/money-currencies' component={SiteExchange} />} />
      </Switch>
    </Router>
  )
}

export default App
