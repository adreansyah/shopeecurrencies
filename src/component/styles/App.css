.align-currencies {
  text-align: center;
  padding: 10px;
  padding-top:20px; 
}

.italic {
  font-style: italic;
}

.el-currencies {
  margin-top: 12px;
  border: 1px solid #ebebeb;
}

.el-flex {
  display: flex;
}

.el-size-left{
  width: 100%;
  text-align: left;
  padding: 6px;
}

.el-fair-size-left{
  width:85%;
  text-align: left
}

.is-list{
  position: initial;
  display: inline-flex;
  /* padding-top: 24px; */
}
.el-size-right {
  border-left: 1px solid #ebebeb;
  background: #f24a58;
  color: white;
  width:14%
}

.header-curr {
  border-bottom: 1px solid #ebebeb;
  margin: 4px 0px 12px 0px;
}

.float-l {
  float: left;
  display: grid;
}

.is-flex{
  display: flex;
  justify-content: flex-end;
  padding: 20px 0px 0px 0px;
  margin-bottom: 6px;
}

.txt-left{
  text-align: left
}

.txt-right{
  text-align: right
}

.txt-center{
  text-align: center
}

.el-number-dictionary {
  font-size: 16px;
  font-weight: bold;
  font-style: italic;
}

.el-not-found{
  margin:30px
}