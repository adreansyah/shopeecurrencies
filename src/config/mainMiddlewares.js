import { combineReducers } from 'redux';
import { combineEpics } from 'redux-observable';

/**
 *
 * Reducers
 */

import flowCurrencies from './reducers/flowCurrency';

/**
 *
 * Reducers
 */

/**
 *
 * Epics
 */
import { epicCurrencies } from './epic/epicCurrency';

/**
 *
 * Epics
 */

export const rootReducer = combineReducers({
  flowCurrencies,
});

export const SetupEpicCombines = combineEpics(epicCurrencies);
