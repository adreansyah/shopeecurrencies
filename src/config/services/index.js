// @flow
import axios from 'axios';

const Axios = () => {
	let instance = axios.create();
	instance.defaults.headers.common['Content-Type'] = 'application/json';
	instance.interceptors.response.use(
		response => {
			return response;
		},
		error => {
			let code,
				status,
				message = '';
			if (error.response) {
				code = error.response.status;
				status = error.response.statusText;
				message = error.response.data.detail
					? error.response.data.detail
					: error.response.data.title
					? error.response.data.title
					: error.response.data.message
					? error.response.data.message
					: error.response.statusText;
			} else {
				code = status = message = error.message;
			}
			let errorData = {
				code,
				status,
				message,
			};
			throw errorData;
		}
	);
	return instance;
};
const execString = (str: string) => {
	return str.replace(matches => matches[1].toUpperCase());
};

class API {
	constructor() {
		this.endpoints = {};
	}

	createEntities(arrayOfEntity) {
		const MultipleEndpoints = arrayOfEntity.join('/');
		const name = execString(MultipleEndpoints);
		this.endpoints['entities'] = this.createBasicCRUDEndpoints({ name });
	}

	createBasicCRUDEndpoints({ name }) {
		let endpoints = {};

		let resourceURL = `https://api.exchangeratesapi.io/${name}`;
		endpoints.getAll = () => Axios().get(resourceURL);

		endpoints.getOne = (config: null) => Axios().get(`${resourceURL}`, config);

		endpoints.create = (config: null) => Axios().post(resourceURL, config);

		endpoints.update = (config: null) => Axios().put(resourceURL, config);

		endpoints.updateById = (id: number, config: null) => axios.put(`${resourceURL}/${id}`, config);

		endpoints.delete = (id: number, config: null) => axios.delete(`${resourceURL}/${id}`, config);

		return endpoints;
	}
}

export const Setup = new API();
