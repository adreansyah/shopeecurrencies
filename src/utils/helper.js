import numeral from 'numeral';
export const removeByAttr = (arr, attr, value) => {
	var i = arr.length;
	while (i--) {
		if (arr[i] && arr[i].hasOwnProperty(attr) && arguments.length > 2 && arr[i][attr] === value) {
			arr.splice(i, 1);
		}
	}
	return arr;
};

export const NUMBERING_SET_FORMAT = '0,0.00';

export const isSetNumber = n => {
	return `${numeral(n).format('0.00')}`;
};

export const isSetThousand = n => {
	return numeral(n).format(`${NUMBERING_SET_FORMAT}`);
};
