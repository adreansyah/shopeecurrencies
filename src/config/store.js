import { createStore, applyMiddleware } from 'redux';
import thunkMiddleware from 'redux-thunk';
import { createLogger } from 'redux-logger';
import { composeWithDevTools } from 'redux-devtools-extension';
import { createEpicMiddleware } from 'redux-observable';
import { rootReducer, SetupEpicCombines } from './mainMiddlewares';
const loggerMiddleware = createLogger();
const epicMiddleware = createEpicMiddleware();
let store = null;
store = createStore(
  rootReducer,
  composeWithDevTools(
    applyMiddleware(thunkMiddleware, epicMiddleware, loggerMiddleware)
  )
);
epicMiddleware.run(SetupEpicCombines);

export default store
