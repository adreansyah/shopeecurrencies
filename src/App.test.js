// @flow

import React from "react";
import { render } from "@testing-library/react";
import { Provider } from "react-redux";
import { shallow, mount } from "enzyme";
import { Setup } from "config/services";
import { isSetThousand, isSetNumber } from "utils/helper";
import store from "config/store";
import App from "./App";
import SiteExchange from "pages/exchangecurrencies";
import flowCurrencies from "config/reducers/flowCurrency";

describe("Show Render Apps ", () => {
  it("renders", () => {
    shallow(
      <Provider store={store}>
        <App />
      </Provider>
    );
  });
});

describe("Show element page notfound  ", () => {
  const sel = id => `[data-test="${id}"]`;
  const hasContentnotfound = wrapper =>
    wrapper.find(sel("not-found")).length === 1;
  it("show value not found", () => {
    const wrapper = mount(
      <Provider store={store}>
        <SiteExchange />
      </Provider>
    );
    expect(hasContentnotfound(wrapper)).toBe(true);
  });
});

describe("Test API Currencies by Region  ", () => {
  it("Define Currency response and function convert", async function() {
    Setup.createEntities(["latest"]);
    const rawResponse = await Setup.endpoints.entities.getOne({
      params: {
        base: "USD"
      }
    });
    const currenciesOption = Object.keys(rawResponse.data.rates);
    const regIDR = "137,625.01";
    const regCHF = "9.76";
    const regBRL = "40.74";
    const regSGD = "13.50";
    const defaultAmount = 10; //you can change this default by USD, as you want to change :)
    expect(rawResponse.data.rates.currenciesOption).toBe(
      rawResponse.data.rates.currenciesOption
    );
    expect(
      isSetThousand(isSetNumber(rawResponse.data.rates.IDR * defaultAmount))
    ).toBe(regIDR);
    expect(
      isSetThousand(isSetNumber(rawResponse.data.rates.CHF * defaultAmount))
    ).toBe(regCHF);
    expect(
      isSetThousand(isSetNumber(rawResponse.data.rates.BRL * defaultAmount))
    ).toBe(regBRL);
    expect(
      isSetThousand(isSetNumber(rawResponse.data.rates.SGD * defaultAmount))
    ).toBe(regSGD);
  });
});

describe("Get Initial States", () => {
  test("is correct", () => {
    const action = { type: "GET_DUMMY_USD_DEFAULT_CURRENCY" };
    const initialState = {
      data: [],
      selectedData: []
    };
    expect(flowCurrencies(undefined, action)).toEqual(initialState);
    expect(flowCurrencies(undefined, action)).toMatchSnapshot();
  });
});
