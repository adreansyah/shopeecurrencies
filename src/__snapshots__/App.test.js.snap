// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Get Initial States is correct 1`] = `
Object {
  "data": Array [],
  "selectedData": Array [],
}
`;
